### DISCUSSION TITLE
Mypy type checker fails after upgrading from 2.1.4 to 2.2.0

### DISCUSSION THREAD

**Discussion by brain-buster** [apache/airflow#19006] [General]
### Apache Airflow version

2.2.0 (latest released)

### Operating System

Ubuntu 20.04.3 LTS

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==2.3.0
apache-airflow-providers-ftp==2.0.1
apache-airflow-providers-http==2.0.1
apache-airflow-providers-imap==2.0.1
apache-airflow-providers-mongo==2.1.0
apache-airflow-providers-postgres==2.3.0
apache-airflow-providers-sqlite==2.0.1

### Deployment

Other 3rd-party Helm chart

### What happened

After updating to Apache Airflow 2.2.0 type checker Mypy completely fails with the following error:

./lib/python3.8/site-packages/airflow/_vendor/connexion/apis/abstract.py:16: note: In module imported here,
./lib/python3.8/site-packages/airflow/_vendor/connexion/apis/__init__.py:1: note: ... from here,
./lib/python3.8/site-packages/airflow/_vendor/connexion/__init__.py:5: note: ... from here,
./lib/python3.8/site-packages/airflow/api_connexion/exceptions.py:21: note: ... from here,
./lib/python3.8/site-packages/airflow/exceptions.py:26: note: ... from here,
./lib/python3.8/site-packages/airflow/__init__.py:63: note: ... from here,
./my_project/airflow_dags/my_file_with_dag_definition.py:5: note: ... from here:
./lib/python3.8/site-packages/airflow/_vendor/connexion/spec.py:169: error: invalid syntax  [syntax]
            spec.setdefault('consumes', ['application/json'])  # type: List[str]

### How to reproduce

Run 'mypy' command in recursively mode within project:

`mypy .`

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
Why and how do you run mypy?  Which version? What are the exclusions you make ? How does it relate to mypy configuration that is part of Airflow itsefl?

Just to explain: Our Mypy is working perfectly fine for airflow - It's part of the static checks for every commit. There is no "guarantee" that all versions of mypy will be running always with all configuration and without exclusions - as usual with any static checkers.

I would love to know what you do and what your configuration you use, because I am genuinly interested, but it is hardly an issue with Airflow - this is a discussion really - that's why I converted it.

**Reply by brain-buster** [reply]
I am using mypy==0.910
I will try to use python3.10 since there are huge different in type hints between Python 3.8 and 3.10 versions.

**Comment by potiuk** [comment]
We are using 0.770 https://github.com/apache/airflow/blob/931bb858833bd0dc88146357f245e907ae404c9b/constraints-3.6.txt#L352 and since we are using namespace packages we had to do some workarounds https://github.com/apache/airflow/blob/main/scripts/in_container/run_mypy.sh 

Also you have to remember that you have to have alll dependende packages installed to make sure that mypy runs in a repeatable fashion. Airflow has > 500 dependent packages and you need to have them installed all via `ci_all` extra wtth the same versions that are "golden" set of dependencies (`pip install apache-airflow[ci_all] --constraint ...` - see https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html ). Only then you might get repeatable `mypy` experience.

**Reply by brain-buster** [reply]
Thank you, will try to downgrade mypy version at first.

**Reply by brain-buster** [reply]
Sorry for noisy discussion, but are you sure that this type comment is properly defined?

spec.setdefault('consumes', ['application/json'])  # type: List[str]

Ref: ./lib/python3.9/site-packages/airflow/_vendor/connexion/spec.py

**Reply by brain-buster** [reply]
`from typing import List` is missed in ./lib/python3.9/site-packages/airflow/_vendor/connexion/spec.py

I tried to remove this type comment and mypy `error: invalid syntax` was resoled

**Comment by brain-buster** [comment]
@potiuk Please, can you reopen issue since this type comment is incorrect?
`spec.setdefault('consumes', ['application/json'])  # type: List[str]`

Should be:
`consumes = spec.setdefault('consumes', ['application/json'])  # type: List[str]`

or completely remove type comment:
`spec.setdefault('consumes', ['application/json'])`

**Reply by potiuk** [reply]
I am on vacation now - but feel free to make PR removing the comment if you think it is invalid. There is no need to re-open issue for that

**Reply by zachliu** [reply]
i'm having the same issue. my findings are more or less the same as @brain-buster:
* airflow 2.2.0 reaches this offending line (previous versions do not) https://github.com/apache/airflow/blob/5dd690b57a20ca944deb8d96e89ec6ae6161afeb/airflow/_vendor/connexion/spec.py#L169 because of this new addition in 2.2.0 https://github.com/apache/airflow/blob/5dd690b57a20ca944deb8d96e89ec6ae6161afeb/airflow/exceptions.py#L26

* as @brain-buster pointed out, using `# type: ...` after a non-assignment statement is no longer allowed after python 3.8
* https://github.com/zalando/connexion (i assume `airflow/_vendor` is based off this) completely removed the type comment after 2.7.0 https://github.com/zalando/connexion/blob/e30971c5f30bc8290e70e242bdb3f056b1274a64/connexion/spec.py#L164

**Reply by brain-buster** [reply]
Exactly!

**Comment by zachliu** [comment]
@brain-buster you beat me to it :grin: 
https://github.com/apache/airflow/pull/19065

**Comment by potiuk** [comment]
Yeah. I see the problem was with running 'mypy` when checking your own 'DAGs' wth mypy. Merged I hope we can cherry-pick it to 2.2.1.
